cpuid/src/common.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::arch::x86_64::CpuidResult;
+use std::arch::x86_64::__cpuid_count;
+use std::arch::x86_64::__get_cpuid_max;
+
+use kvm::CpuId;
+use kvm_bindings::kvm_cpuid_entry2;
+
+const INTEL: &[u8; 12] = b"GenuineIntel";
+const AMD: &[u8; 12] = b"AuthenticAMD";
+
+const EXT_FUNCTION: u32 = 0x80000000;
+
+pub enum Error {
+    InvalidParameters(String),
+    NotSupported,
+    SizeLimitExceeded,
+}
+
+fn get_cpuid(function: u32, count: u32) -> Result<CpuidResult, Error> {
+    unsafe {
+        if __get_cpuid_max(function & EXT_FUNCTION).0 < function {
+            return Err(Error::InvalidParameters(format!(
+                "Function not supported: 0x{:x}",
+                function
+            )));
+        }
+
+        let entry = __cpuid_count(function, count);
+        if entry.eax == 0 && entry.ebx == 0 && entry.ecx == 0 && entry.edx == 0 {
+            return Err(Error::InvalidParameters(format!(
+                "Invalid count: {}",
+                count
+            )));
+        }
+
+        Ok(entry)
+    }
+}
+
+/// Extracts the CPU vendor id from leaf 0x0.
+///
+pub fn get_vendor_id() -> Result<[u8; 12], Error> {
+    match get_cpuid(0, 0) {
+        Ok(vendor_entry) => {
+            let bytes: [u8; 12] = unsafe {
+                std::mem::transmute([vendor_entry.ebx, vendor_entry.edx, vendor_entry.ecx])
+            };
+            Ok(bytes)
+        }
+        Err(_e) => Err(Error::NotSupported),
+    }
+}
+
+/// Replaces the `cpuid` entries corresponding to `function` with the entries from the host's cpuid.
+///
+pub fn emulate_host_cpuid_function(cpuid: &mut CpuId, function: u32) -> Result<(), Error> {
+    // copy all the CpuId entries, except for the ones with the provided function
+    let mut entries: Vec<kvm_cpuid_entry2> = Vec::new();
+    for entry in cpuid.mut_entries_slice().iter() {
+        if entry.function != function {
+            entries.push(*entry);
+        }
+    }
+
+    // add all the host leaves with the provided function
+    let mut count: u32 = 0;
```

Comment:
I'm not sure I understand how you are using `count` here. To my knowledge, it's supposed to be an additional input parameter to `cpuid`, passed in `ECX`. I.e. leaf/function goes in `EAX`, and, if the leaf needs it, a sub-leaf (or count) goes into `ECX` before `cpuid` is executed.

As far as I know, it is also the `index` member of `kvm_cpuid_entry2`. I.e. `entry.function` is the leaf, while `entry.index` is the sub-leaf (or count, or whatever). So `kvm_cpuid_entry2.index` is not (as the name might suggest) the index of that entry into the list, but the value of the sub-leaf for that leaf/function.

Reply:
Ah, yes, it's right. I misread your code - you were only counting sub-leaves. Sorry.